### PR TITLE
[Enhancement] Support large delete only batches for primary key table.

### LINKED ISSUE #4772: Support large delete only batches for primary key table
## Enhancement
Currently, primary key tablet only supports commit multi-segment rowsets if it's all upserts(no deletes), if it's a upsert/delete mixed rowset, we only support single segment.
In the recent PR, delete syntax is added to primary key table, we may encounter large delete batches, to support this scenario, we should support multi-segment delete only rowsets, since it's all delete, it's relatively easy to support.

### PR DESCRIPTION
**Pull request by Linkerist**
## What type of PR is this：
- [ ] bug
- [x] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4772 

## Problem Summary(Required) ：

Supports committing multi-segment rowsets if it's all upserts/deletes, and supports committing a single-segment rowset.

### CODE REVIEW

File: be/src/storage/memtable.h [hunk 1]
```diff
@@ -46,6 +48,8 @@ class MemTable {
 
     bool is_full() const;
 
+    static void init_flush_chunk_state() { state = FlushChunkState::original; }
```
Comment (decster) [inline be/src/storage/memtable.h:51]: why is this static/global?

**Review by decster — CHANGES_REQUESTED** [review]
also add some ut

**Comment by Linkerist** [comment]
do not review for now

File: be/src/storage/memtable.h [hunk 2]
```diff
@@ -18,6 +18,8 @@ class TabletSchema;
 
 namespace vectorized {
 
+enum class FlushChunkState { original, upserts, deletes, mixed };
```
Comment (chaoyli) [inline be/src/storage/memtable.h:21]: What's the original is meaning?

Comment (chaoyli) [reply be/src/storage/memtable.h:21]: unknown is a better name

Comment (Linkerist) [reply be/src/storage/memtable.h]: it mean that there is no flush before, it is first flush.

File: be/src/storage/memtable.h (hunk 1 above)
Comment (Linkerist) [reply be/src/storage/memtable.h]: changed.

File: be/src/storage/delta_writer.cpp [hunk 3]
```diff
@@ -355,4 +413,14 @@ const char* DeltaWriter::_state_name(State state) const {
...
 }
 
+void DeltaWriter::calc_flush_chunk_state(bool upserts, bool deletes) {
+    if (upserts && !deletes) {
+        this->_curr_flush_chunk_state = FlushChunkState::upserts;
+    } else if (!upserts && deletes) {
+        this->_curr_flush_chunk_state = FlushChunkState::deletes;
+    } else {
+        this->_curr_flush_chunk_state = FlushChunkState::mixed;
+    }
```
Comment (decster) [inline be/src/storage/delta_writer.cpp:423]: what if both are false

File: be/src/storage/delta_writer.h [hunk 4]
```diff
@@ -23,6 +23,8 @@ class MemTable;
 
 enum WriteType { LOAD = 1, LOAD_DELETE = 2, DELETE = 3 };
 
+enum class FlushChunkState { unknown, upserts, deletes, mixed };
```
Comment (decster) [inline be/src/storage/delta_writer.h:26]: enum uses upper case by convention

File: be/src/storage/delta_writer.h [hunk 5]
```diff
@@ -123,6 +127,8 @@ class DeltaWriter {
     const TabletSchema* _tablet_schema;
 
     std::unique_ptr<FlushToken> _flush_token;
+    FlushChunkState _flush_chunk_state = FlushChunkState::unknown;
+    FlushChunkState _curr_flush_chunk_state = FlushChunkState::unknown;
```
Comment (decster) [inline be/src/storage/delta_writer.h:131]: It seems putting state in rowset writer rather than delta_writer can simply things a lot

**Review by decster — DISMISSED** [review]

File: be/src/storage/memtable.h (hunk 2 above)
Comment (chaoyli) [reply be/src/storage/memtable.h]: Original seems to have no correspondence to no flush.

Comment (Linkerist) [reply be/src/storage/memtable.h]: good idea, changed

File: be/src/storage/delta_writer.h (hunk 5 above)
Comment (Linkerist) [reply be/src/storage/delta_writer.h]: good idea, changed.

File: be/src/storage/delta_writer.h (hunk 4 above)
Comment (Linkerist) [reply be/src/storage/delta_writer.h]: sorry i forgot it, changed.

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/memtable.cpp [hunk 6]
```diff
@@ -222,10 +222,10 @@ Status MemTable::flush() {
     int64_t duration_ns = 0;
     {
         SCOPED_RAW_TIMER(&duration_ns);
-        if (!_deletes || _deletes->empty()) {
-            RETURN_IF_ERROR(_rowset_writer->flush_chunk(*_result_chunk));
-        } else {
+        if (_deletes) {
             RETURN_IF_ERROR(_rowset_writer->flush_chunk_with_deletes(*_result_chunk, *_deletes));
+        } else {
+            RETURN_IF_ERROR(_rowset_writer->flush_chunk(*_result_chunk));
```
Comment (decster) [inline be/src/storage/memtable.cpp:228]: flush_chunk should also maintain _flush_chunk_state

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**